### DISCUSSION TITLE
Postgres bulk update

### DISCUSSION THREAD

**Discussion by gboor** [sqlalchemy/sqlalchemy#11768] [Usage Questions] [answered]
I am using SQLAlchemy with the ORM to handle data in a Postgres database.
I need to update thousands of rows in one go and SQLAlchemy has support for bulk updates, but it's not doing what I want it to do;

```
stmt = update(MyModel).values(name=bindparam("name"))
db.execute(stmt, [{"id": 1, "name": "blah1"}, {"id": 2, "name": "blah2"}, {"id": 3, "name": "blah3"}])
```

Does work, but actually generates 3 separate UPDATE queries in a transaction;

```
BEGIN (implicit);
UPDATE table SET name = "blah1" WHERE id = 1;
UPDATE table SET name = "blah2" WHERE id = 2;
UPDATE table SET name = "blah3" WHERE id = 3;
COMMIT;
```

This doesn't really scale well. Postgres has support for bulk updates using the UPDATE ... FROM structure;

```
UPDATE table
SET name = myvalues.name
FROM (
  VALUES
    (1, 'blah1'),
    (2, 'blah2'),
    (3, 'blah3')
) AS myvalues (id, name)
WHERE table.id = myvalues.id
```

Is there any way to generate such a query using SQLAlchemy?

**Comment by zzzeek** [comment, marked as answer]
you need to use the [Values](https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.values) construct explicitly for this non-standard syntax

```
from sqlalchemy import column
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy import table
from sqlalchemy import values

t= table('t', column('id', Integer), column('name', String))

v = values(
    column('id', Integer),
    column('name', String),
    name="my_values"
).data(
    [(1, 'name1'), (2, 'name2'), (3, 'name3')]
)

stmt = t.update().values(name=v.c.name)

print(stmt)
```

**Reply by CaselIT** [reply]
maybe it could be documented in the update docs? interesting that it somehow generates a `FROM`

**Reply by zzzeek** [reply]
I think values() could use some docs because what's there doesnt really say much

UPDATE..FROM is in the tutorial here: https://docs.sqlalchemy.org/en/20/tutorial/data_update.html#update-from  it's got a full section

**Reply by CaselIT** [reply]
that section misses the values example. Also is seems that also mssql has a from in the update https://learn.microsoft.com/en-us/sql/t-sql/queries/update-transact-sql?view=sql-server-ver16

**Reply by gboor** [reply]
Thanks for the follow-up. The construction with the values works to generate the query I want!

**Reply by CaselIT** [reply]
added docs here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5986

**Comment by umayrh** [comment]
@zzzeek I'm using `2.0.36`, and `stmt = t.update().values(name=v.c.name)` doesn't generate the  `WHERE table.id = myvalues.id` clause.

**Reply by zzzeek** [reply]
use [update.where](https://docs.sqlalchemy.org/en/20/core/dml.html#sqlalchemy.sql.expression.Update.where) to add a WHERE clause to an UPDATE

**Comment by Adamantish** [comment]
I'd be interested if someone who knows SQLAlchemy internals well could explain why the postgres extension couldn't be extended further to make this how `flush()` works by default? Done in batches of some size.

**Reply by CaselIT** [reply]
mainly because it would require rewriting a lot of code. same reason it doesn't use a similar approach in oracle

**Comment by Apakottur** [comment]
I'm trying to use the `UPDATE..FROM` pattern from @zzzeek's comment with `Mapped` based column definitions:
```python
from sqlalchemy import update
from sqlalchemy import values

from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column


class DbBase(DeclarativeBase):
    pass


class MyTable(DbBase):
    __tablename__ = "my_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]


update_values = values(
    MyTable.id,
    MyTable.name,
    name="update_values",
).data([(1, "Alice"), (2, "Bob")])

query = (
    update(MyTable)
    .values(
        {
            MyTable.name: update_values.c.name,
        }
    )
    .where(MyTable.id == update_values.c.id)
)
```

This works correctly and generates the correct DB query:
```sql
UPDATE
    my_table
SET
    name = update_values.name
FROM (
    VALUES (1, 'Alice'),
        (2, 'Bob')) AS update_values (id, name)
WHERE
    my_table.id = update_values.id
```

However, running `mypy` gives::
```text
a.py:19: error: Argument 1 to "values" has incompatible type "InstrumentedAttribute[int]"; expected "ColumnClause[Any]"  [arg-type]
a.py:20: error: Argument 2 to "values" has incompatible type "InstrumentedAttribute[str]"; expected "ColumnClause[Any]"  [arg-type]
```

Is this a mypy issue or am I doing something wrong?
I am passing the table columns directly to `.values()`.

**Reply by CaselIT** [reply]
the issue seems to be on the type side (so the mypy error is wrong)

**Reply by Apakottur** [reply]
Should I then open an issue on this repo?

btw Pyright gives the same error:
```text
a.py:19:5 - error: Argument of type "InstrumentedAttribute[int]" cannot be assigned to parameter "columns" of type "ColumnClause[Any]" in function "values"
    "InstrumentedAttribute[int]" is not assignable to "ColumnClause[Any]" (reportArgumentType)
a.py:20:5 - error: Argument of type "InstrumentedAttribute[str]" cannot be assigned to parameter "columns" of type "ColumnClause[Any]" in function "values"
    "InstrumentedAttribute[str]" is not assignable to "ColumnClause[Any]" (reportArgumentType)
```

**Reply by CaselIT** [reply]
if you can provide a fix feel free to just open a PR, otherwise we could have an issue for it, sure

**Comment by AlexKovyazin** [comment]
In case someone has such a problem.
The documentation [has an example](https://docs.sqlalchemy.org/en/20/tutorial/data_update.html#update-from) of updating from values, but in the sqlalchemy 1 style.

I tried this and everything works fine:
```python
async def bulk_update_value_field(self, rows: Sequence[PydanticSchema]) -> None:
    values = (
        Values(MyTableOrm.user_id, MyTableOrm.value_field, name="values")
        .data([(r.user_id, r.value_field) for r in rows])
    )
    stmt = (
        update(MyTableOrm)
        .where(MyTableOrm.user_id == values.c.user_id)
        .values(value_field=values.c.value_field)
    )
    await self.session.execute(stmt)
```
It generates this query:
```sql
   UPDATE users_profiles
      SET value_field=values.value_field
     FROM (VALUES ('id1', 'value1'),
                                ('id2', 'value2')
           ) AS values (user_id, value_field)
    WHERE my_table_name.user_id = values.user_id
RETURNING my_table_name.id;
```
